README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# hyetor

[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/kvantas/hyetor?branch=master&svg=true)](https://ci.appveyor.com/project/kvantas/hyetor)
[![Travis build
status](https://travis-ci.org/kvantas/hyetor.svg?branch=master)](https://travis-ci.org/kvantas/hyetor)
[![Coverage
status](https://codecov.io/gh/kvantas/hyetor/branch/master/graph/badge.svg)](https://codecov.io/github/kvantas/hyetor?branch=master)
[![DOI](https://zenodo.org/badge/145962234.svg)](https://zenodo.org/badge/latestdoi/145962234)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

`hyetor` (Vantas 2018) is an R package (R Core Team 2018) that provides
a collection of tools that analyze fixed interval precipitation records.
It can be used to:

1.  Aggregate precipitation values.
2.  Split precipitation time series to independent rainstorms using
    predefined monthly maximum dry period duration of no precipitation
    (Restrepo-Posada and Eagleson 1982).
3.  Compile Unitless Cumulative Hyetographs (Bonta 2004).
4.  Find maximum rainfall intensities.
5.  Categorize rainstorms using Huff’s quartile classification (Huff
    1967).
6.  Calculate rainfall erosivity values using a number of energy
    equations (Wischmeier and Smith 1958; Brown and Foster 1987;
    McGregor et al. 1995).
7.  Create missing values summaries.

For more details checkout the package’s
[website](https://kvantas.github.io/hyetor/) and the vignettes:

## Installation

You can install the development version from Github with:

``` r
# install.packages("devtools")
devtools::install_github("kvantas/hyetor")
```

## Using hyetor

The functions that are provided by `hyetor` are:

  - Functions that can be used to preprocess precipitation time-series:
    `hyet_create`, `hyet_fill`, `hyet_aggregate` and `hyet_split`.
  - Functions that analyze precipitation time-series: `hyet_erosivity`,
    `hyet_intensities`, `hyet_missing` and `hyet_uch`.

The data sets that are provided by `hyetor` are:

  - `prec5min`, time series that comes from the weather station ‘Arna’
    in Greece. The owner of that weather station is the Ministry of
    Environment and Energy. The time series period is from 12/1954 to
    05/1956 and the time-step is 5 minutes.

## Example

This is a minimal example which shows how to use the package’s functions
to analyze the internal data set and compute erosivity values.

Load libraries and view data:

``` r
library(hyetor)
library(tibble)
library(dplyr)
library(lubridate)

prec5min
#> # A tibble: 48,209 x 2
#>    date                 prec
#>    <dttm>              <dbl>
#>  1 1954-12-14 07:40:00     0
#>  2 1954-12-14 07:45:00     0
#>  3 1954-12-14 07:50:00     0
#>  4 1954-12-14 07:55:00     0
#>  5 1954-12-14 08:00:00     0
#>  6 1954-12-14 08:05:00     0
#>  7 1954-12-14 08:10:00     0
#>  8 1954-12-14 08:15:00     0
#>  9 1954-12-14 08:20:00     0
#> 10 1954-12-14 08:25:00     0
#> # ... with 48,199 more rows
```

Compute the missing values ratio per year:

``` r
prec5min %>%
  hyet_fill(time_step = 5, ts_unit = "mins") %>%
  hyet_missing() %>%
  group_by(year) %>%
  summarise(na_ratio = mean(na_ratio))
#> # A tibble: 3 x 2
#>    year na_ratio
#>   <dbl>    <dbl>
#> 1  1954    0.493
#> 2  1955    0.723
#> 3  1956    0.619
```

Split to independent rainstorms and calculate erosivity values with
cumulative precipitation greater than 12,7 mm:

``` r
prec5min %>%
  hyet_split(time_step = 5, ts_unit = "mins") %>%
  hyet_erosivity(time_step = 5) %>%
  ungroup() %>%
  filter(cum_prec > 12.7) %>%
  select(begin, duration, cum_prec, erosivity)
#> # A tibble: 29 x 4
#>    begin               duration    cum_prec erosivity
#>    <dttm>              <time>         <dbl>     <dbl>
#>  1 1955-03-03 10:05:00 " 780 mins"     13.0      7.33
#>  2 1955-04-14 14:15:00 " 620 mins"     38.7    130.  
#>  3 1955-04-15 11:15:00 1235 mins       24.7     30.4 
#>  4 1955-05-11 13:30:00 " 375 mins"     28.2    123.  
#>  5 1955-07-15 14:50:00 " 125 mins"     19.4    162.  
#>  6 1955-08-30 14:30:00 " 150 mins"     23.8    115.  
#>  7 1955-09-02 12:05:00 " 360 mins"     25.2    124.  
#>  8 1955-09-04 13:10:00 " 170 mins"     20.4     52.5 
#>  9 1955-09-28 18:05:00 " 100 mins"     35.6    366.  
#> 10 1955-10-01 01:50:00 1115 mins       29.6     47.4 
#> # ... with 19 more rows
```

## Meta

  - Bug reports, suggestions, and code are welcome. Please see
    [Contributing](/CONTRIBUTING.md).
  - Licence:
      - All code is licensed MIT.
      - All data are from the public data sources in
        <http://www.hydroscope.gr/>.
  - To cite `hyetor` please
use:

<!-- end list -->

``` 
  Vantas, (2018). hyetor: R package to analyze fixed interval precipitation time series, URL: https://kvantas.github.io/hyetor/,
  DOI:http://doi.org/10.5281/zenodo.1403156
```

A BibTeX entry for LaTeX users is

    @Manual{ vantas2018hyetor,
        author = {Konstantinos Vantas},
        title = {{hyetor}: R package  to analyze fixed interval precipitation time series},
        doi = {http://doi.org/10.5281/zenodo.1403156},
        year = {2018},
        note = {R package},
        url = {https://kvantas.github.io/hyetor/},
      }

## References

<div id="refs" class="references">

<div id="ref-bonta2004development">

Bonta, JV. 2004. “Development and Utility of Huff Curves for
Disaggregating Precipitation Amounts.” *Applied Engineering in
Agriculture* 20 (5). American Society of Agricultural; Biological
Engineers: 641.

</div>

<div id="ref-brown1987storm">

Brown, LC, and GR Foster. 1987. “Storm Erosivity Using Idealized
Intensity Distributions.” *Transactions of the ASAE* 30 (2). American
Society of Agricultural; Biological Engineers: 379–0386.

</div>

<div id="ref-huff1967time">

Huff, Floyd A. 1967. “Time Distribution of Rainfall in Heavy Storms.”
*Water Resources Research* 3 (4). Wiley Online Library: 1007–19.

</div>

<div id="ref-McGregor1995">

McGregor, K.C., Ron Bingner, A.J. Bowie, and G.R. Foster. 1995.
“Erosivity Index Values for Northern Mississippi” 38 (January):
1039–47.

</div>

<div id="ref-CRAN">

R Core Team. 2018. *R: A Language and Environment for Statistical
Computing*. Vienna, Austria: R Foundation for Statistical Computing.

</div>

<div id="ref-restrepo1982identification">

Restrepo-Posada, Pedro J, and Peter S Eagleson. 1982. “Identification of
Independent Rainstorms.” *Journal of Hydrology* 55 (1-4). Elsevier:
303–19.

</div>

<div id="ref-vantas2018hyetor">

Vantas, Konstantinos. 2018. *hyetor: R Package to Analyze Fixed Interval
Precipitation Time Series*.
<https://doi.org/http://doi.org/10.5281/zenodo.1403156>.

</div>

<div id="ref-wischmeier1958rainfall">

Wischmeier, Walter H, and Dwight D Smith. 1958. “Rainfall Energy and Its
Relationship to Soil Loss.” *EOS, Transactions American Geophysical
Union* 39 (2). Wiley Online Library: 285–91.

</div>

</div>


NEWS.md
# hyetor 0.0.0.9000

* Initial version


R/utils.R
#' Return error if not positive integer
#'
#' @noRd
count_check <- function(x, x_name) {
  if (!assertthat::is.count(x)) {
    stop(paste(x_name, "must be a positive integer"), call. = FALSE)
  }
}

#' Return error if x not a vector of values
#'
#' @noRd
vector_check <- function(time_step) {
  if (!(time_step %in% c(5, 10, 15, 30))) {
    stop(
      "`time_step` must have one of the values 1, 5, 10, 15 or 30 minutes.",
      call. = FALSE
    )
  }
}

#' return error if units are valid
#'
#' @noRd
units_check <- function(x, minhour = TRUE) {
  if (minhour) {
    if (!(x %in% c("mins", "hours"))) {
      stop(paste("units must be either `mins` or `hours`"), call. = FALSE)
    }
  } else {
    if (!(x %in% c("mins", "hours", "days", "months", "quarter", "year"))) {
      stop(paste(
        "units must be on of `mins`, `hours`, `days`, `months`,",
        "`quarter` or `year`"
      ), call. = FALSE)
    }
  }
}

#' return error if crit_dur is not a numeric vector of 12 values
#' @noRd
crit_dur_check <- function(x) {
  if (!is.numeric(x) | (length(x) != 12) | any(x < 0)) {
    stop(paste("crit_dur must be a numeric vector of 12 values"),
      call. = FALSE
    )
  }
}

#' return error if not valid energyy equation
#'
#' @noRd
en_eq_check <- function(x) {
  valid_equat <- c("brown_foster", "mcgregor_mutch", "wisch_smith")
  error_msg <- paste(
    "`en_equation` must be on of:",
    "`brown_foster`, `mcgregor_mutch`, `wisch_smith`"
  )
  if (!(x %in% valid_equat)) {
    stop(error_msg, call. = FALSE)
  }
}

#' Check for a valid hyetograph
#' @noRd
hyet_check <- function(hyet) {
  # check for expected names and values
  suppressWarnings(
    if (!(tibble::is.tibble(hyet) & (all(c("date", "prec") %in% names(hyet))) &
      lubridate::is.POSIXct(hyet$date) & is.numeric(hyet$prec))) {
      error_msg <- paste(
        "Error: `hyet` is not a valid hyetograph. Please use",
        "function `hyet_create`."
      )
      stop(error_msg, call. = FALSE)
    }
  )
}

#' Fill hyetographs
#'
#' @noRd
util_fill <- function(hyet, time_step, ts_unit) {

  # create an empty time series
  empty_ts <- tibble::tibble(date = seq(
    from = min(hyet$date, na.rm = TRUE),
    to = max(hyet$date, na.rm = TRUE),
    by = paste(time_step, ts_unit)
  ))

  # merge time series
  dplyr::left_join(empty_ts, hyet, by = "date")
}

#' Utility function for hyetograph aggregation
#' @noRd
util_aggr <- function(hyet, time_step, ts_unit = "mins") {

  # create aggregated date
  hyet$date <- lubridate::ceiling_date(
    hyet$date,
    paste(time_step, ts_unit)
  )

  # group by date
  hyet <- dplyr::group_by(hyet, .data$date)

  # summarise values
  dplyr::summarise(hyet, prec = sum(.data$prec))
}

#' Utility function for hyetographs rolling sum
#'
#' @noRd
util_roll_sum <- function(hyet, win_size) {
  hyet$prec <- RcppRoll::roll_sum(hyet$prec,
    n = win_size, fill = 0,
    align = "left", na.rm = TRUE
  )
  hyet
}


#' Calculate intensities using an hyetograph
#'
#' @noRd
rain_intensities <- function(hyet, time_step, ts_unit) {
  from_dur <- paste(time_step, ts_unit)
  ts_dur <- lubridate::duration(time_step, ts_unit)
  tibble::tibble(
    start = hyet$date[1],
    end = tail(hyet$date, 1),
    duration = difftime(tail(hyet$date, 1) + ts_dur, hyet$date[1],
      units = "hours"
    ),
    cum_prec = sum(hyet$prec, na.rm = TRUE),
    quartile = huff_class(c(0, cumsum(na.omit(hyet$prec)))),
    int_mean = .data$cum_prec / as.numeric(.data$duration),
    int_5min = max_roll_sum(hyet, from_dur, "5 mins") * 12,
    int_10min = max_roll_sum(hyet, from_dur, "10 mins") * 6,
    int_15min = max_roll_sum(hyet, from_dur, "15 mins") * 4,
    int_30min = max_roll_sum(hyet, from_dur, "30 mins") * 2,
    int_1hr = max_roll_sum(hyet, from_dur, "1 hours"),
    int_3hr = max_roll_sum(hyet, from_dur, "3 hours") / 3,
    int_6hr = max_roll_sum(hyet, from_dur, "6 hours") / 6,
    int_12hr = max_roll_sum(hyet, from_dur, "12 hours") / 12,
    int_24hr = max_roll_sum(hyet, from_dur, "1 days") / 24,
    int_48hr = max_roll_sum(hyet, from_dur, "2 days") / 48
  )
}

#' Find maximum rolling window sum
#'
#' @noRd
max_roll_sum <- function(hyet, from_dur, to_dur) {

  # compute rolling window
  win <- lubridate::as.duration(to_dur) / lubridate::as.duration(from_dur)
  res <- NA

  # if window < 1 return NA
  if (win < 1) {
    return(res)
  }

  # do not calculate rolling sum if window = 1
  if (win == 1) {
    res <- (max(hyet$prec))
  }
  else if (assertthat::is.count(win)) {
    tmp <- util_roll_sum(hyet, win)
    res <- suppressWarnings(max(tmp$prec, na.rm = TRUE))
  }

  ifelse(is.infinite(res), NA, res)
}
